draft-ietf-quic-transport.md

```diff
@@ -870,44 +870,107 @@ different IP or port at either endpoint, due to NAT rebinding or mobility, as
+
+The server receives this packet and determines whether it potentially creates a
+new connection (see {{packet-handling}}).  If the packet might generate a new
+connection, the server then checks whether it understands the version that the
+client has selected.
+
+If the packet contains a version that is acceptable to the server, the server
+proceeds with the handshake ({{handshake}}).  This commits the server to the
+version that the client selected.
 
-When the server receives a packet from a client with the long header format, it
-compares the client's version to the versions it supports.
+
+### Sending Version Negotiation Packets {#send-vn}
 
 If the version selected by the client is not acceptable to the server, the
-server discards the incoming packet and responds with a Version Negotiation
-packet ({{packet-version}}).  This includes a list of versions that the server
-will accept.
+server responds with a Version Negotiation packet ({{packet-version}}).  This
+includes a list of versions that the server will accept.
 
-To avoid packet amplification attacks a server MUST NOT send a Version
-Negotiation packet that is larger than the packet it responds to.  It is
-anticipated that this is ample space for all QUIC versions that a single server
-might need to advertise.
+A server sends a Version Negotiation packet for any packet with an unaccepable
+version if that packet could create a new connection.  This allows a server to
+process packets with unsupported versions without retaining state.  Though
+either the Client Initial packet or the version negotiation packet that is sent
+in response could be lost, the client will send new packets until it
+successfully receives a response or it abandons the connection attempt.
 
-A server sends a Version Negotiation packet for every packet that it receives
-with an unacceptable version.  This allows a server to process packets with
-unsupported versions without retaining state.  Though either the initial client
-packet or the version negotiation packet that is sent in response could be lost,
-the client will send new packets until it successfully receives a response.
 
-If the packet contains a version that is acceptable to the server, the server
-proceeds with the handshake ({{handshake}}).  This commits the server to the
-version that the client selected.
+### Handling Version Negotiation Packets {#handle-vn}
 
-When the client receives a Version Negotiation packet from the server, it should
-select an acceptable protocol version.  If the server lists an acceptable
-version, the client selects that version and reattempts to create a connection
-using that version.  Though the contents of a packet might not change in
+When the client receives a Version Negotiation packet, it first checks that the
+packet number and connection ID match the values it sent in a Client Initial
+packet.  If this check fails, the packet MUST be discarded.
```

Comment:
Another potential timing issue:  Client sends Client Initial + 0-RTT data.  Client Initial is lost in transit.  Version Negotiation comes back responding to 0-RTT packets because the server only received those.  Per the current text, it MUST ignore the server's VN packets until it has retransmitted the CI and gotten *that* VN packet.  Hopefully rare, since if you're doing 0-RTT you're unlikely to use an unsupported version -- but not impossible.

Maybe we should simply say "match the values it sent in a packet on this connection"?